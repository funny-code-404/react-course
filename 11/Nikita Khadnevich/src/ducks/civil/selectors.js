
export const Civildata = (state) => state.civil.data.civilizations;
export const CivildataDetail = (state) => state.civil.dataDetail;
export const Civilerror = (state) => state.civil.error;
export const Civilurl = (state) => state.civil.url;
export const CivilisFetching = (state) => state.civil.isFetching;
export const Civilpath = (state) => state.civil.path;

