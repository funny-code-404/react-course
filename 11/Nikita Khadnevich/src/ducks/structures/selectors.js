export const StructuresData = (state) => state.structures.dataSTRU.structures;
export const StructuresDataDetail = (state) => state.structures.dataSTRUDetail;
export const StructuresError = (state) => state.structures.error;
export const StructuresUrl = (state) => state.structures.url;
export const StructuresisFetching = (state) => state.structures.isFetching;
export const StructuresPath = (state) => state.structures.path;