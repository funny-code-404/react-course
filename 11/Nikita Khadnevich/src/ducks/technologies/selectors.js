export const TechData = (state) => state.technologies.dataTech.technologies;
export const TechDataDetail = (state) => state.technologies.dataTechDetail;
export const TechError = (state) => state.technologies.error;
export const TechUrl = (state) => state.technologies.url;
export const TechisFetching = (state) => state.technologies.isFetching;
export const TechPath = (state) => state.technologies.path;