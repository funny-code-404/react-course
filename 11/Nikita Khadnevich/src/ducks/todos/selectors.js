export const todosSelector = (state) => state.todos.data;
