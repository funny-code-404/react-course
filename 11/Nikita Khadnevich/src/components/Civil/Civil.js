import React, { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { ACTION_GET_CIVIL_Requested } from '../../ducks/civil/actions';
import { Civildata } from '../../ducks/civil/selectors'
import  { baseUrl, Urlpath } from '../Api/Api'
import { indicator, ListAge, Sort } from '../SmallElems/SmallElems'


const Civil = (props) => {
  const { civilizations } = Urlpath
  const data = useSelector(Civildata);
  const dispatches = useDispatch();
  const { civil } = indicator;
  const propsUrl = props.match.url

  const getFetch = (url, path, arr) => {
    if (!arr) {
      dispatches(ACTION_GET_CIVIL_Requested(`${url}/${path}`));
    } else return null
  }

  useEffect(() => {
    getFetch(baseUrl, civilizations, data)
  }, []);

  return (
    <>
      <ListAge data={data} blockName={civil} elem={civilizations} propsUrl={propsUrl} Sort={Sort}/>
    </>
  )
};

export default Civil

