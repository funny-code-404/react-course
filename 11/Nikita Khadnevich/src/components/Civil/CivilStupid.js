import React from 'react'
import {BrowserRouter as Router, Switch, Route, Link} from 'react-router-dom'
import { indicator, ButtonClose } from '../SmallElems/SmallElems'
const { civil, civInfo, civDetailinfo, civilStupid } = indicator
import { Civilerror } from '../../ducks/civil/selectors'


export const CivilInfoStupid = (props) => {
   const { children: [item, urlCiv, handleclick, civilizations] } = props

   return (
      <>
         <p key={item.name+civilStupid}>Имя: {item.name}</p>
         <p key={item.expansion+civilStupid}>Класс война: {item.expansion}</p>
         <p key={item.army_type+civilStupid}>Тип Армии: {item.army_type}</p>
         <p key={item.team_bonus+civilStupid}>Командный бонус: {item.team_bonus}</p>

         {(item.unique_unit && item.unique_unit[0]) ? item.unique_unit.map((linkUnit,i) => {
         return (
            <div key={'CivilsItem'+civilStupid+i} className={civilizations+'Item'}>
               <Link to={`${urlCiv}/unique_unit`} id='unique_unit' onClick={handleclick} data-path={linkUnit}>{i+1}. Уникальный юнит {item.name}</Link>
            </div>
            )
            }) : 
            <div key={'CivilsItemUnit'+civilStupid} className={civilizations+'NoItem'}>
               <p>У цивилизации нет уникальных юнитов</p>
            </div>
         }

         { (item.unique_tech && item.unique_tech[0]) ? item.unique_tech.map((linkTech,i) => {
               return (
               <div key={'CivilsItem2'+civilStupid+i} className={civilizations+'Item2'}>
                  <Link to={`${urlCiv}/unique_tech`} id='unique_tech' onClick={handleclick} data-path={linkTech}>{i+1}. Уникальная технология {item.name}</Link>
               </div>
               )
            }) :  
            <div key={'CivilsItemNoTech'+civilStupid} className={civilizations+'NoItem'}>
               <p>У цивилизации нет уникальных технологий</p>
            </div>
         }
      </>
   )
};

export const CivilUnitDetailStupid = (props) => {
   const { children: [$, handleLocation] } = props
   const { civDetailinfo } = indicator
   return (
      <>
      <ul key={'Skils'+civDetailinfo} id='Skils'>
         <ButtonClose handleLocation={handleLocation} idName={'Skils'} indicator={civDetailinfo} selector={Civilerror} />
         <li>Имя: {$.name}</li>
         <li >Фракция: {$.age}</li>
         <li >Броня: {$.armor}</li>
         <li >Атака: {$.attack}</li>
         <li id='liHit'>Единицы доровья: {$.hit_points}</li>
         <li >Скорость перемещения: {$.movement_rate}</li>
         <li >Время перезарядки: {$.reload_time}</li>
         <li >Дальность видимости: {$.line_of_sight}</li>
         {  $.attack_bonus ? <ul key={'AttackBonus'+civDetailinfo} className='AttackBonus'> 
            <li >Бонус к атаке:</li>
            {$.attack_bonus && $.attack_bonus.map((item, i) => {
               return (
                  <li key={item+civDetailinfo}>{item}</li>
                  )
               })
            }
         </ul> : null }
      </ul>
   </>
   )
}

export const CivilTechDetailStupid = (props) => {
   const { children: [$, handleLocation] } = props
   const { civDetailinfo } = indicator
   return (
      <>
         <ul key={'Skils'+civDetailinfo} id='Skils'>
            <ButtonClose handleLocation={handleLocation} idName={'Skils'} indicator={civDetailinfo} selector={Civilerror} />
            <li>Имя: {$.name}</li>
            <li>Фракция: {$.age}</li>
            <li>Описание: {$.description}</li>
            <li>Атака: {$.expansion}</li>
            { $.cost.Food || $.cost.Gold ? <li>Стоимость постройки: {`${$.cost.Food} еды`} , {`${$.cost.Gold} золота`}</li> : null }
            <li>Время постройки: {$.build_time}</li>
         </ul>
      </>
   )
}

export default CivilInfoStupid