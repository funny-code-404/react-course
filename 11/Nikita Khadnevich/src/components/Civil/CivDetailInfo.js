import React, { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useHistory }  from 'react-router-dom'
import { CivildataDetail, CivilisFetching, Civilpath, Civilerror } from '../../ducks/civil/selectors'
import { ACTION_GET_CIVIL_DETAIL_Succeed, ACTION_GET_CIVIL_DETAIL_FAILED } from '../../ducks/civil/actions'
import { ButtonClose, indicator} from '../SmallElems/SmallElems'
import { CivilUnitDetailStupid, CivilTechDetailStupid } from './CivilStupid'


const CivDetailInfo = (props) => {
   const history = useHistory();
   const $ = useSelector(CivildataDetail);
   const fetchStatus = useSelector(CivilisFetching);
   const path = useSelector(Civilpath)
   const  dispatches = useDispatch()
   const { civDetailinfo } = indicator
   
   const handleLocation = () => {
      history.go(-1);
      dispatches(ACTION_GET_CIVIL_DETAIL_Succeed({}))
      dispatches(ACTION_GET_CIVIL_DETAIL_FAILED(null))

   }
   // Сделать дата мап б проверить есть ли поля с unique_unit и unique_tech

   return (
      <>
         {
            ($.id && !fetchStatus && path == 'unique_unit') ? 
               <CivilUnitDetailStupid>
                  {$}{handleLocation}
               </CivilUnitDetailStupid>
            : ($.id && !fetchStatus && path == 'unique_tech') ?
               <CivilTechDetailStupid>
                  {$}{handleLocation}
               </CivilTechDetailStupid>
            : (!$.id && !fetchStatus) ?
               <ButtonClose handleLocation={handleLocation} idName={'Skils'} indicator={civDetailinfo} selector={Civilerror}/>
            : null
         }
      </>
   )
}

export default CivDetailInfo
