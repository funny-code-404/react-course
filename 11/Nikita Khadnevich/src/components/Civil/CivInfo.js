import React, { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import {BrowserRouter as Router, Switch, Route, useHistory, useParams} from 'react-router-dom'
import { ACTION_GET_CIVIL_Requested, ACTION_GET_CIVIL_REQUESTED_DETAIL , ACTION_GET_ROUTE_CIVIL_MENU, ACTION_GET_CIVIL_DETAIL_FAILED } from '../../ducks/civil/actions';
import { Civildata } from '../../ducks/civil/selectors'
import CivDetailInfo from './CivDetailInfo'
import  CivilInfoStupid  from './CivilStupid'
import  { baseUrl, Urlpath } from '../Api/Api'
import { ButtonGoBack, indicator } from '../SmallElems/SmallElems'


const CivInfo = (props) => {
   const history = useHistory()
   const params = useParams()
   const data = useSelector(Civildata)
   const  dispatches = useDispatch()
   const { civInfo } = indicator
   const { civilizations } = Urlpath
   const urlCiv = props.match.url

   const getFetch = (url, path, arr) => {
      if (!arr) {
      dispatches(ACTION_GET_CIVIL_Requested(`${url}/${path}`));
      } else return null
   }
   useEffect(() => {
   getFetch(baseUrl, civilizations, data)
   }, []);

   const handleclick = (e) => {
      const targetPath = e.target.dataset.path
      const targetId = e.target.id
      dispatches(ACTION_GET_ROUTE_CIVIL_MENU(targetId));
      targetPath ?
      dispatches(ACTION_GET_CIVIL_REQUESTED_DETAIL(targetPath)) : null
   }
         //^^^^
         //Сначала была такая логика
         // const targetUnit = data[e.target.id-1].unique_unit
         // const targetTech = data[e.target.id-1].unique_tech
         // (targetPath == 'unique_tech' && targetTech.length > 0) ?
         // dispatches(ACTION_GET_CIVIL_REQUESTED_DETAIL(targetPath)) : null
         // dispatches(ACTION_GET_CIVIL_DETAIL_FAILED('Данный юнит не существует'))
         // где id = это порядковый номер элемента массива
         //Затем я заменил на CivilInfoStupid map и потребность в диспатче еррора отпала, ибо если нет unique, то она просто не рисует кнопки
         
   //1. Пройтись по масиву и забрать уникальное значение
   //2ю задиспатчить его уже в имеющийся новый массив

   // вставлю линку даты и создам новый стор для юниов 
   // путем распаршивания ее через джоин(например)
   // а уже потом у меня будет линковаться все как нужно
   const handleLocation = () => {
      history.push('/civilizations');
   }

   return ( 
   <>
      {data && data.map((item, i) => {
         if (params.id === (item.name+item.id)) {
            return (
               <div key={'Wrapper'+civInfo}>
                  <ButtonGoBack handleLocation={handleLocation} idName={'goback'+civInfo} indicator={civInfo}/>
                  <div key={'Items'+civInfo+i} className={'items'+civInfo}>
                     <CivilInfoStupid key={'stupid'+civInfo+i}> 
                        {item}{urlCiv}{handleclick}{civilizations}
                     </CivilInfoStupid>
                  </div>
                  <Switch>
                     <Route path={`${urlCiv}/:id`} component={CivDetailInfo} />
                  </Switch>
               </div>
            )
         } 
      })}
   </>
   )
}

export default CivInfo


