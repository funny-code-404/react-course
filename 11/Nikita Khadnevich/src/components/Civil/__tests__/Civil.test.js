import React from 'react';
import Civil from '../Civil';
import { Civildata } from '../../../ducks/civil/selectors'
import { configure, shallow, mount } from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
configure({ adapter: new Adapter() });
import configureMockStore from 'redux-mock-store' //ES6 modules

jest.mock('../../../ducks/civil/selectors')
const mockFnDispatch = jest.fn();
const mockFnEffect = jest.fn();
const mockFnSelector = jest.fn()

jest.mock('react-redux', () => ({
   useSelector: () => mockFnSelector(),
   useDispatch: () => mockFnDispatch,
   useEffect: () => mockFnEffect
}));

describe('<Civil /> ', () => {
   const mockFn = jest.fn()
   const props = {
      match: {
         url : 'Make some url'
      },
      onClick: mockFn,
      text: 'Bye',
   };

   let wrapper;
   beforeEach(() => {
      wrapper = mount(<Civil {...props} propsUrl={props.match.url}/>)
   });

   it('Is Civil component', () => {
      expect(wrapper).toHaveLength(1);
   })

   //-----------------------------------------------//

   it('Is data from store', () => {
      Civildata.mockReturnValue()
      expect(wrapper).toHaveLength(1)
   })

   //-----------------------------------------------//

   it('should are working Methods in Civil', () => {
      const middlewares = [] //middlewares are empty
      const mockStore = configureMockStore(middlewares)// create Stor
      const ACTION_GET = () => ({ type: 'ACTION_GET' })// create Action
      const initialState = { data: [1, 2, 3, 4] }// Init StorContain
      const store = mockStore(initialState) // Put contain into Stor
      const civil = store.getState().data //useSelector
      const props = { match: { url: 'some http...' }} // add Props
   

      const getFetch = (url, path, res) => {
         if ( Boolean(res) == false) {
            mockFnDispatch(ACTION_GET(`${url}${path}`))
         } else null
      }
      mockFnEffect(getFetch( 'www.tut.by', '/avto', civil))

      // console.log(`civil -- `, store.getState().data)
      // console.log(`props -- `, props.match.url)

      const wrapper = mount(
         // <Provider store={store}>
            <Civil data={civil} getFetch={getFetch} useEffect={mockFnEffect} {...props}/>
         //* </Provider> */}
      );
      expect(wrapper).toMatchSnapshot();
   })

   //-----------------------------------------------//

   it('should to find div element in the DOM', () => {
      expect(wrapper.find('div')).toHaveLength(1);// Проверка на то, что есть элемент с className = CivilizationsWrapper
   })

   //-----------------------------------------------//

   it('should to find CivilizationsWrapper in the DOM', () => {
      expect(wrapper.find('div.civilizationsWrapper')).toHaveLength(1);// Проверка на то, что есть элемент с className = CivilizationsWrapper
   })

   //--------------------------------------------//

   it('should set props correctly', () => {
      // console.log(`wrapper.props().text`, wrapper.props().text)
      const nextText = 'Hello';
      wrapper.setProps( {text: nextText});
      expect(wrapper.props().text).toEqual(nextText);
      // console.log(`wrapper.props().text`, wrapper.props().text)
   })
})

//функция shallow используется для того, чтобы замаунтить компонент(return)