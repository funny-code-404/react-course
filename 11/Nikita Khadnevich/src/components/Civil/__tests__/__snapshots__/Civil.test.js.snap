// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Civil />  should are working Methods in Civil 1`] = `
<Civil
  data={
    Array [
      1,
      2,
      3,
      4,
    ]
  }
  getFetch={[Function]}
  match={
    Object {
      "url": "some http...",
    }
  }
  useEffect={
    [MockFunction] {
      "calls": Array [
        Array [
          undefined,
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
>
  <ListAge
    blockName="_civil"
    elem="civilizations"
    propsUrl="some http..."
  >
    <div
      className="civilizationsWrapper"
      key="civilizationsWrapper"
    />
  </ListAge>
</Civil>
`;
